src/librustdoc/test.rs

```diff
@@ -473,10 +473,15 @@ pub fn make_test(s: &str,
     if dont_insert_main || already_has_main {
         prog.push_str(everything_else);
     } else {
-        prog.push_str("fn main() {\n");
+        let returns_result = everything_else.trim_end().ends_with("(())");
+        let (main_pre, main_post) = if returns_result {
+            ("fn main() { fn _inner() -> Result<(), impl std::fmt::Debug> {",
```

Comment:
Yes, but there's still an error: It should be `core::fmt::Debug`. That said, `fn main() -> Result<..>` doesn't work on `#![no_std]` code for now, because the `Termination` trait is only available on `std`.